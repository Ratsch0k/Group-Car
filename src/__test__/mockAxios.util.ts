jest.mock('axios');
import axios from 'axios';
const mockedAxios = axios as jest.Mocked<typeof axios>;

mockedAxios.head.mockImplementationOnce((path: string) => {
  if (path === '/auth') {
    return Promise.resolve({
      headers: {
        'xsrf-token': 'TEST TOKEN',
      },
    });
  }
});

export default mockedAxios;
