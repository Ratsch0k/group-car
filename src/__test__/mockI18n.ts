jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
}));
