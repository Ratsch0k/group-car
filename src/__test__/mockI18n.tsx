import React from 'react';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
  Trans: (props: any) => {
    return <div>{props.i18nKey}</div>;
  },
}));
