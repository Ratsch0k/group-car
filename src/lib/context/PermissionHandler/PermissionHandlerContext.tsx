import {createContext} from 'react';

export const Permission = ['geolocation'] as const;


export interface PermissionHandlerContext {
  checkPermission: (name: typeof Permission[number]) =>
  Promise<PermissionStatus>;

  requestPermission: (name: typeof Permission[number]) =>
  Promise<void>;
}

export const PermissionHandlerContext =
createContext<PermissionHandlerContext>({
  checkPermission: () => Promise.reject(new Error('Context is missing')),
  requestPermission: () => Promise.reject(new Error('Context is missing')),
});

export default PermissionHandlerContext;
