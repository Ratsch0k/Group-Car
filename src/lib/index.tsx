export * from './components';
export * from './ModalRouter';
export * from './theme';
export * from './requests';
export * from './Routes';
export * from './context';
export * from './hooks';
