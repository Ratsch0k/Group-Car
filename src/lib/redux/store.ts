import {configureStore, getDefaultMiddleware} from '@reduxjs/toolkit';
import rootReducers from './slices';
import {
  routerMiddleware,
} from 'connected-react-router';
import history from 'lib/redux/history';
import * as Sentry from '@sentry/react';

const sentryReduxEnhancer = Sentry.createReduxEnhancer({
  actionTransformer: (action) => {
    return action;
  },
  stateTransformer: (state) => {
    return state;
  },
});


/**
 * Create redux store.
 */
const store = configureStore({
  reducer: rootReducers(history),
  devTools: true,
  middleware: getDefaultMiddleware().concat(routerMiddleware(history)),
  enhancers: [sentryReduxEnhancer],
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
