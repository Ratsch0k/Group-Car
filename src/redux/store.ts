import {configureStore} from '@reduxjs/toolkit';
import rootReducer from './slices';

/**
 * Create redux store.
 */
const store = configureStore({
  reducer: rootReducer,
  devTools: true,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
