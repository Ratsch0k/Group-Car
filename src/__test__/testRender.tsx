import React from 'react';
import {render, RenderResult} from '@testing-library/react';
import {ConnectedRouter} from 'connected-react-router';
import {Provider} from 'react-redux';
import history from '../lib/redux/history';
import {StylesProvider, ThemeProvider} from '@material-ui/styles';
import testTheme from './testTheme';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import thunk from 'redux-thunk';
import {RootState} from '../lib/redux/store';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';


export interface TestRenderResult<T> {
  mockStore: MockStoreCreator<T>;
  store: MockStoreEnhanced<unknown>;
}

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

/**
 * Testing generateClassName function to
 * avoid weird fails due to random class names
 * @param rule Rule
 * @param stylesheet Stylesheet
 * @returns The class name
 */
const generateClassName = (rule, stylesheet) => {
  return `${stylesheet.options.classNamePrefix}-${rule.key}`;
};


/**
 * Render function for testing purposes.
 * Provides the necessary base components and
 * contexts.
 * @param children The children to render
 * @returns The result of the render function.
 */
function testRender<T>(state: T, children: React.ReactNode):
RenderResult & TestRenderResult<T> {
  const middleware = [thunk];
  const mockStore = configureMockStore<T>(middleware);
  const initialState = {
    ...defaultState,
    ...state,
  };
  const store = mockStore(initialState);

  const renderResult = render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={testTheme}>
          <StylesProvider generateClassName={generateClassName}>
            {children}
          </StylesProvider>
        </ThemeProvider>
      </ConnectedRouter>
    </Provider>,
  );

  return {
    ...renderResult,
    mockStore,
    store,
  };
}

export default testRender;
