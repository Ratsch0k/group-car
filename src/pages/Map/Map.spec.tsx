import '../../__test__/mockAxios';
import { fireEvent, waitFor } from "@testing-library/react";
import React from 'react';
import { LatLng, Map } from "leaflet";
import { MapContext } from "../../lib";
import MapComponent from './Map';
import testRender from '../../__test__/testRender';
import { RootState } from "../../lib/redux/store";
import {act} from "react-dom/test-utils";
import PermissionHandlerContext from '../../lib/context/PermissionHandler/PermissionHandlerContext';

const mockWatchPosition = jest.fn();
jest.mock('../../lib/hooks/useGeolocation', () => ({
  __esModule: true,
  default: () => ({
    watchPosition: mockWatchPosition,
  }),
}));

describe('Map', () => {
  let permissionContext: PermissionHandlerContext;

  beforeEach(() => {
    permissionContext = {
      checkPermission: jest.fn().mockResolvedValue({
        state: 'granted',
      }),
      requestPermission: jest.fn().mockResolvedValue(undefined),
    };
  });

  afterEach(() => {
    mockWatchPosition.mockReset();
    jest.unmock('../../lib/hooks/useGeolocation');
  })

  const customRender = (state: Partial<RootState>, mapContext: MapContext) => {
    return testRender(
      state,
      <MapContext.Provider value={mapContext}>
        <PermissionHandlerContext.Provider value={permissionContext}>
          <MapComponent />
        </PermissionHandlerContext.Provider>
      </MapContext.Provider>
    );
  };

  
  it('watches position with geolocation', async () => {
    const map = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      flyTo: jest.fn(),
    } as unknown as Map;

    mockWatchPosition.mockResolvedValue(undefined);

    const mapContext = {
      map,
    } as MapContext;

    const state = {
      group: {
        selectedGroup: null,
        ids: [],
        entities: {},
        loading: false,
      },
    };

    const {baseElement} = customRender(state, mapContext);

    fireEvent.click(baseElement.querySelector('#show-current-position')!);
    await waitFor(() => expect(mockWatchPosition).toHaveBeenCalledTimes(1));
    expect(mockWatchPosition).toHaveBeenCalledWith(expect.any(Function));
  });


  describe('if a car is selected and selection is not disabled', () => {
    it('registers click event listener on map if a car is selected and select', async () => {
      const map = {
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        flyTo: jest.fn(),
      } as unknown as Map;

      const mapContext = {
        map,
        selectedCar: {
          groupId: 1,
          carId: 1,
          name: 'car',
          driverId: 1,
          Driver: {
            id: 1,
            username: 'driver'
          },
        },
        selectionDisabled: false,
      } as MapContext;

      const geolocation = {
        watchPosition: jest.fn(),
      } as unknown as Geolocation;
      (navigator.geolocation as any) = geolocation;

      const state = {
        group: {
          selectedGroup: null,
          ids: [],
          entities: {},
          loading: false,
        },
      };

      customRender(state, mapContext);

      await waitFor(() => expect(map.addEventListener).toHaveBeenCalledTimes(1));
      expect(map.addEventListener).toHaveBeenCalledWith('click', expect.any(Function));
    });
  });
});