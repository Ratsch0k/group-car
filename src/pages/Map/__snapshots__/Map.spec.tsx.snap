// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map zooms to current position when first found 1`] = `
<body>
  <div>
    <div
      class="makeStyles-map leaflet-container leaflet-grab leaflet-touch-drag"
      style="position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="https://a.tile.openstreetmap.org/6/33/21.png"
                style="width: 256px; height: 256px; left: -202px; top: -122px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        />
        <div
          class="leaflet-pane leaflet-marker-pane"
        >
          <img
            alt=""
            class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
            src="marker_pos.svg"
            style="margin-left: -12px; margin-top: -12px; width: 24px; height: 24px; left: -94px; top: 59px; z-index: 59;"
            tabindex="0"
          />
        </div>
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        >
          <div
            class="leaflet-control-zoom leaflet-bar leaflet-control"
          >
            <a
              aria-label="Zoom in"
              class="leaflet-control-zoom-in"
              href="#"
              role="button"
              title="Zoom in"
            >
              +
            </a>
            <a
              aria-label="Zoom out"
              class="leaflet-control-zoom-out"
              href="#"
              role="button"
              title="Zoom out"
            >
              −
            </a>
          </div>
        </div>
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JS library for interactive maps"
            >
              Leaflet
            </a>
             | © 
            <a
              href="http://osm.org/copyright"
            >
              OpenStreetMap
            </a>
             contributors
          </div>
        </div>
      </div>
      0
    </div>
  </div>
</body>
`;
